### DISCUSSION TITLE
Monitoring convergence?

### DISCUSSION THREAD

**Discussion by olibclarke** [zhonge/cryodrgn#165] [Q&A]
Hi,

What is the best/appropriate way to monitor convergence during `train_vae`? Is monitoring the learning curve useful, or is there a better metric?

I am trying `train_vae` on a small 20k subset of EMPIAR-11043, downsampled to 128px, just to get a feel for how it works. In an initial round, a small but well defined cluster of outlier particles was eliminated using a Z score threshold. In a second round using the remaining particles, the UMAP plot seems somewhat featureless (although there is genuine variability between the kmeans clusters, broadly similar to what we see using classification without alignments). 

I am wondering what parameters to tweak to optimize this? Obviously using more particles will help (probably a lot!), but are there any other parameters to consider in such a case? Number of iterations I guess, anything else?

<img width="614" alt="image" src="https://user-images.githubusercontent.com/19766818/197772684-87395052-8bb4-4b4a-89ac-6786d316cab5.png">

<img width="588" alt="image" src="https://user-images.githubusercontent.com/19766818/197775074-e12f33de-f6ef-4e48-af2a-11bd890eadb3.png">

**EDIT: I am trying on a larger set of 200k images, and the UMAP already looks much more "featured" and interpretable, even at epoch 8, so I guess particle number makes a big difference (you can also see the small junk population very clearly here):**
<img width="593" alt="image" src="https://user-images.githubusercontent.com/19766818/197828302-9bf79286-0d03-476d-967a-bc075f41a5eb.png">

**EDIT2: And here it is with the full dataset (700k images), after 15 epochs:**
![umap_hexbin](https://user-images.githubusercontent.com/19766818/198013747-f20f6153-36a9-470b-80fc-74eb1dc0c5e9.png)

**Comment by Guillawme** [comment]
Hello,

This is discussed in this preprint: https://doi.org/10.1101/2022.08.09.503342 

But in brief, you can assess convergence by looking at these things:

- the learning curve levels off to some asymptote: this tells about the entire network
- the UMAP plot stops changing between epochs (the shapes and relative positions of discrete clusters might still change slightly, but the number of clusters and the relative populations of particles in each cluster should stabilize after enough epochs): this tells about the encoder network, since the Z values analyzed by UMAP are the output of the encoder
- the maps you get from the cluster centroids also stabilize and stop changing much between epochs: this tells about the decoder network, since it is the one that generates maps

I don't know how many particles is too few, but I did get meaningful results with just over 40k particles in a previous project (see https://doi.org/10.7554/eLife.71420 ).

I hope this helps!

**Reply by olibclarke** [reply]
Thanks very much @Guillawme that is very helpful! Just reading the preprint now and it is very helpful also. Is there any script to calculate the per iteration UMAP plots all at once, or graphs of derived parameters? E.g. the latent embedding shifts described in the preprint?

Re the number of particles, I am running a test now using the same parameters with a 200k subset, to see how that changes things.

**Reply by Guillawme** [reply]
I don't have any automation for this: I simply use `cryodrgn analyze` on a few epochs starting from the last one I ran, and check how the UMAP plots change. This and the learning curve are often enough to decide which epoch to use for interpretation, but sometimes you need to look at the maps too.

I am also interested to know if there is a tool to generate the other convergence plots shown in the preprint.

**Reply by olibclarke** [reply]
Actually in the preprint it is suggested that such a script exists - on page 5 it is stated, in reference to the discussed convergence heuristics summarized in Box 1, "_We provide a dedicated script within cryoDRGN that automatically calculates and plots these criteria._" @zhonge which script does this refer to?

**Reply by olibclarke** [reply]
Ah! Reading further in the preprint there is a script "analyze_convergence.py" which looks to be part of `cryodrgn_utils` - however I can't see it as an option in the version I have - maybe in an unreleased branch?
<img width="865" alt="image" src="https://user-images.githubusercontent.com/19766818/198039037-a58e39ec-e84b-473d-9895-a9adfab33328.png">

**Reply by olibclarke** [reply]
Ah I see - it is there, in the cryodrgn utils directory, as a standalone script, just not as part of the command line package yet

<img width="358" alt="image" src="https://user-images.githubusercontent.com/19766818/198039730-c9670f2b-fd4c-4619-8b28-9425625fbad2.png">

**Reply by olibclarke** [reply]
And it works just as described! 😄 
<img width="1040" alt="image" src="https://user-images.githubusercontent.com/19766818/198044711-ee366556-ff77-4d50-9cb3-845339a73854.png">
